### PR TITLE
Added deterministic push order test per connection

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Added test

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/xds/pushqueue.go [hunk 1]
```diff
@@ -56,12 +50,6 @@ func (p *PushQueue) Enqueue(con *Connection, pushRequest *model.PushRequest) {
 		return
 	}
 
-	// If its already in progress, merge the info and return
-	if request, f := p.processing[con]; f {
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/pushqueue.go:60]: This can not happen.  The uint tests have all passed after removing the code.

Comment (ramaraochavali) [reply pilot/pkg/xds/pushqueue.go:60]: Would this even create problem if we merge in progress PushRequest as we do not know in which stage it is ? Not sure why it was added in the first place?

**Comment by ramaraochavali** [comment]
LGTM.  But would leave it for @howardjohn for review.

**Review by howardjohn — CHANGES_REQUESTED** [review]
I don't see why this is safe to remove and will not cause a regression. See https://github.com/istio/istio/pull/16211 for more details about this.

If our tests our passing we should improve the tests, not remove the code IMO.

**Comment by hzxuzhonghu** [comment]
@howardjohn After a close looking at the code, the processings is no use at all. 

```
Add To Queue X
Dequeue X
Start push 1 for X
Add To Queue X
Dequeue X
Attempt to start push 2 for X, blocked because only one concurrent push allowed
Add To Queue X
Dequeue X
Attempt to start push 3 for X, blocked because only one concurrent push allowed
......
Push 1 done
Push 3 starts (order is arbitrary! it may not be push 2)
Push 3 finishes
Push 2 starts
Push 2 finishes
```

The order could not change, because when push 1 is running, the push 2 and 3 requests will be merged. After push1 done the merged request will be processed.

**Comment by howardjohn** [comment]
Push 2 and 3 will not merge since we dequeue 2 before we add 3,that is the problem. We end up not merging

**Comment by hzxuzhonghu** [comment]
I know the issue, but looks we could make it more readable

**Comment by howardjohn** [comment]
It may be more readable but the performance is really bad.. see the discussion/graphs in https://github.com/istio/istio/issues/14782 and https://snapshot.raintank.io/dashboard/snapshot/ymgZuKL1hGK1Sj6eK8UqZZI8GtTJNl1R?orgId=2 (the after part is WITHOUT withDone - you can see memory goes extremely high)

**Comment by hzxuzhonghu** [comment]
added test for deterministic order. It is not easy to guarantee the push order without large change of the framework now

**Comment by hzxuzhonghu** [comment]
@howardjohn updated

**Review by howardjohn — APPROVED** [review]

**Pull request merged**